### PR TITLE
Makefile improvements

### LINKED ISSUE #202: [Epic] Makefile generation improvements
- [ ] Use file targets (see [awsx Makefile](https://github.com/pulumi/pulumi-awsx/blob/master/Makefile) for reference). This should allow local development loops to be faster.
- [ ] Generalise the makefile options to not assume providers are owned by Pulumi.
- [ ] Add provider shim mods target (i.e. run `go mod tidy` in a target)
- [ ] Make `provider-ci` consumable externally – something like `npx pulumi-gen-build`

### PR DESCRIPTION
**Pull request by danielrbradley**
Fixes #202 

Create a new version of makefile template: "bridged-v2" which must be opted into. We can enable this on a per-repo basis as we gain confidence in testing.

This includes a few general makefile formatting improvements:
- Explicitly set the "default" target so it's always first.
- Sort targets by name, grouped by type (alias/phony/file).

New makefile changes:

- Use file targets so we can have up-to-date builds.
- Model dependencies so dependencies automatically get built first.
- Use local pulumictl version.
- Use sentinel files to atomically execute targets.

When being rolled out - this will require some minor changes to the build process.

### CODE REVIEW

**Review by thomas11 — APPROVED** [review]
Tricky to review but looks good at first glance. Thank you, this will be a nice improvement!

File: provider-ci/src/make.ts [hunk 1]
```diff
@@ -96,17 +97,57 @@ function phonyTarget(targets: Target[]): Target | undefined {
...
+}
+
+function descendentTargets(target: Target): Target[] {
+  if (target.dependencies == undefined) {
+    return [target];
+  }
+  return deduplicateTargets([
+    target,
+    ...target.dependencies.flatMap((t) =>
+      typeof t != "string" ? descendentTargets(t) : []
```
Comment (lblackstone) [inline provider-ci/src/make.ts]: Should we use `!==` here instead? The `typeof` check on 119 uses `===`.

File: provider-ci/src/make.ts [hunk 2]
```diff
@@ -96,17 +97,57 @@ function phonyTarget(targets: Target[]): Target | undefined {
...
+    ),
+  ]);
+}
+
+function sortTargets(targets: Target[], defaultTarget?: Target | string) {
+  const defaultName =
+    typeof defaultTarget === "string" ? defaultTarget : defaultTarget?.name;
+  const sorted = [...targets].sort((a, b) => {
```
Comment (lblackstone) [inline provider-ci/src/make.ts]: Why is the `[...targets]` needed? Isn't `targets` already an array?

File: provider-ci/src/make.ts [hunk 3]
```diff
@@ -96,17 +97,57 @@ function phonyTarget(targets: Target[]): Target | undefined {
...
+    if (a.name == defaultName) {
```
Comment (lblackstone) [inline provider-ci/src/make.ts]: Should we use `===` here?

**Review by lblackstone — APPROVED** [review]

File: provider-ci/src/make.ts (hunk 2 above)
Comment (danielrbradley) [reply provider-ci/src/make.ts]: Calling `targets.sort` would modify the array passed into the function. I always avoid mutating input arguments unless it's performance critical, so I create a copy of the array then, sort the copy and return the copy.

**Comment by ringods** [comment]
@danielrbradley can we also test drive this at Pulumiverse side for one of the bridged providers?

**Comment by danielrbradley** [comment]
@ringods maybe - we've not had any time to do the work to make this easy to consume outside our core provider base, but feel free to fork for now or find another way to utilize it.

**Pull request merged**